0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+
+The second thing which was a bit challenging was how to model data which is only known at a later point in time and how to model data which not all `Swaps` have in common.
+We decided to go for an _extension table_ approach which allows us to add additional information where needed:
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `role`, `alpha_protocol`, `beta_protocol` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/han_swap_protocols.puml&fmt=svg[HanProtocols]
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
+
+Some notable sub-tables:
+
+* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+One could optimize here and combine all ledger specific identities into an own table `han_bitcoin_identity` to hold all Bitcoin `refund` and `redeem` identities.
+This would mean though that we would need an additional unique id in those sub-tables, i.e. we can't use the `fk` as `pk`.
+
+* `han_bitcoin/ethereum_ledger_state`: is a new table which we introduced in this database schema.
```

Comment:
Can we, by any chance, have a rendering of the model that excludes these tables? I am not keen on thinking about them now because I would delay introducing them until a later point!

Reply:
I would leave the model here for the sake of completeness but can add an excerpt of it without the sate? That'd be straight forward.

Reply:
Yes please, I would just want to look at it without the state (as I think we are not gonna go there yet?)

Reply:
Why is it a separate table in the first place?

Reply:
Ah ok I see, the idea is to have the information related to the state in such table. Something that could be added to the diagram.

Reply:
on it.

Reply:
@thomaseizinger : should be nicer now.